src/metabase/analytics/prometheus.clj

```diff
@@ -0,0 +1,248 @@
+                            :maxPoolSize 15}}
+  Becomes {:numConnections [{:name :numConnections,
+                             :value 15.0, ;; values are all doubles
+                             :timestamp 1662563931039,
+                             :label \"metabase-postgres-app-db\"}
+                            {:name :numConnections,
+                             :value 2.0,
+                             :timestamp 1662563931039,
+                             :label \"db-2-postgres-clean\"}]
+          ...}"
+  [raw-stats]
+  (let [now    (.toEpochMilli (java.time.Instant/now))
+        sample (fn sample [[db-label k v]]
+                 {:name      k
+                  :value     (double v)
+                  :timestamp now
+                  :label     db-label})]
+    (->> raw-stats
+         (mapcat (fn [[db-label values]]
+                   (map (fn [[k v]] [db-label k v]) values)))
+         (map sample)
+         (group-by :name))))
+
+(def ^:private label-translation
+  {:maxPoolSize        {:label       "c3p0_max_pool_size"
+                        :description "C3P0 Max pool size"}
+   :minPoolSize        {:label       "c3p0_min_pool_size"
+                        :description "C3P0 Minimum pool size"}
+   :numConnections     {:label       "c3p0_num_connections"
+                        :description "C3P0 Number of connections"}
+   :numIdleConnections {:label       "c3p0_num_idle_connections"
+                        :description "C3P0 Number of idle connections"}
+   :numBusyConnections {:label       "c3p0_num_busy_connections"
+                        :description "C3P0 Number of busy connections"}})
+
+(defn- ->array
+  "Return an array."
+  ^ArrayList
+  [^java.util.Collection coll]
+  (ArrayList. coll))
+
+(defn- stats->prometheus
+  "Create an ArrayList of GaugeMetricFamily objects containing measurements from the c3p0 stats. Stats are grouped by
+  the property and the database information is attached as a label to multiple measurements of `:numConnections`."
+  [stats]
+  (transduce
+   (comp (filter (fn [[raw-label _measurements]]
+                   (or (label-translation raw-label)
+                       (log/warn (trs "Unrecognized measurement {0} in prometheus stats"
+                                      raw-label)))))
+         (map (fn [[raw-label measurements]]
+                (let [{gauge-label :label desc :description} (label-translation raw-label)
+                      gauge (GaugeMetricFamily.
+                             ^String gauge-label
+                             ^String desc
+                             (->array ["database"]))]
+                  (doseq [m measurements]
+                    (.addMetric gauge
+                                (->array [(:label m)])
+                                (:value m)))
+                  gauge))))
+   (completing conj!
+               (fn [metric-family-samples]
+                 (->array (persistent! metric-family-samples))))
+   stats))
```

Comment:
I wonder if this function would be more readable as a loop adding elements to the mutable `ArrayList`. We are in the Java world anyway...

Reply:
```clojure
(defn stats->prometheus'
  [stats]
  (loop [[stat & remaining] stats
         arr                (ArrayList. (count stats))]
    (let [[raw-label measurements]               stat
          {gauge-label :label desc :description} (label-translation raw-label)]
      (if gauge-label
        (let [gauge (GaugeMetricFamily.
                     ^String gauge-label
                     ^String desc
                     (->array ["database"]))]
          (doseq [m measurements]
            (.addMetric gauge
                        (->array [(:label m)])
                        (:value m)))
          (.add arr gauge)
          (recur remaining arr))
        (recur remaining arr)))))
```
It's a bit shorter but I don't find it particularly easier to read. If you have a strong opinion I'll use it but I'm not seeing a great increase in readability.

Reply:
I had something like the following in mind (sorry, haven't tested):

```clojure
(defn stats->prometheus''
  [stats]
  (let [arr (ArrayList. (count stats))]
    (doseq [[raw-label measurements] stats
            :let [{gauge-label :label desc :description} (label-translation raw-label)]]
      (when gauge-label
        (let [gauge (GaugeMetricFamily.
                     ^String gauge-label
                     ^String desc
                     (List/of "database"))]
          (doseq [m measurements]
            (.addMetric gauge
                        (List/of (:label m))
                        (:value m)))
          (.add arr gauge))))
    arr))
```